src/lib/Actor.hs

```diff
@@ -1,178 +1,67 @@
-   cleanup _ _      (Proc NoTrap linked  _) = void $ forkIO $ killThread linked
 
-asPChan :: BackChan (Msg a) -> PChan a
-asPChan (BackChan _ chan) = PChan (writeChan chan . NormalMsg)
+import Util (onFst, onSnd)
 
-asErrPChan :: BackChan (Msg a) -> PChan (Proc, String)
-asErrPChan (BackChan _ chan) = PChan (writeChan chan . uncurry ErrMsg)
+-- Micro-actors.
 
-send :: MonadActor msg' m => PChan msg -> msg -> m ()
-send p x = liftIO (sendFromIO p x)
+-- In this model, an "actor" is just an IO computation (presumably
+-- running on its own Haskell thread) that receives messages on a
+-- Control.Concurrent.Chan channel.  The idea is that the actor thread
+-- only receives information (or synchronization) from other threads
+-- through messages sent on that one channel, and no other thread
+-- reads messages from that channel.
 
-sendFromIO :: PChan msg -> msg -> IO ()
-sendFromIO = sendPChan
+type Actor a = Chan a -> IO ()
 
-myChan :: MonadActor msg m => m (PChan msg)
-myChan = do cfg <- actorCfg
-            return $ asPChan (selfChan cfg)
-
--- TODO: make a construct to receive in a loop to avoid repeated linear search
-receiveErrF :: MonadActor msg m => (Msg msg -> Maybe a) -> m a
-receiveErrF filterFn = do
-  cfg <- actorCfg
-  liftIO $ find (selfChan cfg) []
-  where find chan skipped = do
-          x <- readBackChan chan
-          case filterFn x of
-            Just y -> do pushBackChan chan (reverse skipped)
-                         return y
-            Nothing -> find chan (x:skipped)
-
-receiveErr :: MonadActor msg m => m (Msg msg)
-receiveErr = receiveErrF Just
-
-receiveF :: MonadActor msg m => (msg -> Maybe a) -> m a
-receiveF filterFn = receiveErrF (skipErr >=> filterFn)
-  where skipErr msg = case msg of NormalMsg x -> Just x
-                                  ErrMsg _ _  -> Nothing
-
-receive :: MonadActor msg m => m msg
-receive = receiveF Just
-
-newBackChan :: IO (BackChan a)
-newBackChan = liftM2 BackChan (newIORef []) newChan
-
-readBackChan :: BackChan a -> IO a
-readBackChan (BackChan ptr chan) = do xs <- readIORef ptr
-                                      case xs of []     -> readChan chan
-                                                 x:rest -> do writeIORef ptr rest
-                                                              return x
-
-pushBackChan :: BackChan a -> [a] -> IO ()
-pushBackChan (BackChan ptr _) xs = do xs' <- readIORef ptr
-                                      writeIORef ptr (xs ++ xs')
-
-class (MonadIO m, Monad m) => MonadActor msg m | m -> msg where
-  actorCfg :: m (ActorConfig msg)
-
-instance MonadActor msg (Actor msg) where
-  actorCfg = Actor ask
-
-instance MonadActor msg m => MonadActor msg (StateT s m) where
-  actorCfg = lift actorCfg
+-- We also define a send-only channel type, to help ourselves not
+-- accidentally read from channels we aren't supposed to.
+newtype PChan a = PChan { sendPChan :: a -> IO () }
```

Comment:
Why is a send only channel called a `PChan`? What's the `P` prefix for?

Reply:
Don't know; I kept it from the previous Actor module.  Partial channel?  @dougalm ?